src/main/scala/nl/hugo/redbook/ch7/Nonblocking.scala

```diff
@@ -0,0 +1,181 @@
+            case Right(b) =>
+              if (ar.isDefined) eval(es)(cb(f(ar.get, b)))
+              else br = Some(b)
+          }
+          p(es)(a => combiner ! Left(a))
+          p2(es)(b => combiner ! Right(b))
+        }
+      }
+
+    // specialized version of `map`
+    def map[A, B](p: Par[A])(f: A => B): Par[B] =
+      es => new Future[B] {
+        def apply(cb: B => Unit): Unit =
+          p(es)(a => eval(es) { cb(f(a)) })
+      }
+
+    def lazyUnit[A](a: => A): Par[A] =
+      fork(unit(a))
+
+    def asyncF[A, B](f: A => B): A => Par[B] =
+      a => lazyUnit(f(a))
+
+    def sequenceRight[A](as: List[Par[A]]): Par[List[A]] =
+      as match {
+        case Nil => unit(Nil)
+        case h :: t => map2(h, fork(sequence(t)))(_ :: _)
+      }
+
+    def sequenceBalanced[A](as: IndexedSeq[Par[A]]): Par[IndexedSeq[A]] = fork {
+      if (as.isEmpty) unit(Vector())
+      else if (as.length == 1) map(as.head)(a => Vector(a))
+      else {
+        val (l, r) = as.splitAt(as.length / 2)
+        map2(sequenceBalanced(l), sequenceBalanced(r))(_ ++ _)
+      }
+    }
+
+    def sequence[A](as: List[Par[A]]): Par[List[A]] =
+      map(sequenceBalanced(as.toIndexedSeq))(_.toList)
+
+    // exercise answers
+
+    /*
+     * We can implement `choice` as a new primitive.
+     *
+     * `p(es)(result => ...)` for some `ExecutorService`, `es`, and
+     * some `Par`, `p`, is the idiom for running `p`, and registering
+     * a callback to be invoked when its result is available. The
+     * result will be bound to `result` in the function passed to
+     * `p(es)`.
+     *
+     * If you find this code difficult to follow, you may want to
+     * write down the type of each subexpression and follow the types
+     * through the implementation. What is the type of `p(es)`? What
+     * about `t(es)`? What about `t(es)(cb)`?
+     */
+    def choice[A](p: Par[Boolean])(t: Par[A], f: Par[A]): Par[A] =
+      es => new Future[A] {
+        def apply(cb: A => Unit): Unit =
+          p(es) { b =>
+            if (b) eval(es) { t(es)(cb) }
+            else eval(es) { f(es)(cb) }
+          }
+      }
+
+    // Exercise 7.11
+    def choiceN[A](p: Par[Int])(ps: List[Par[A]]): Par[A] = ???
+
+    // Exercise 7.11
+    def choiceViaChoiceN[A](a: Par[Boolean])(ifTrue: Par[A], ifFalse: Par[A]): Par[A] =
+      ???
+
+    // Exercise 7.12
+    def choiceMap[K, V](p: Par[K])(ps: Map[K, Par[V]]): Par[V] =
+      ???
+
+    // Exercise 7.13
+    // see `Nonblocking.scala` answers file. This function is usually called something else!
+    def chooser[A, B](p: Par[A])(f: A => Par[B]): Par[B] =
+      ???
+
+    // Exercise 7.13
+    // Note[AD]: I've swapped the arguments back to (t: Par[A], f: Par[A]) they were
+    // (f: Par[A], t: Par[A]) in the original..
+    def choiceViaChooser[A](p: Par[Boolean])(t: Par[A], f: Par[A]): Par[A] =
+      ???
+
+    // Exercise 7.13
+    def choiceNChooser[A](p: Par[Int])(choices: List[Par[A]]): Par[A] =
```

Comment:
Annoyingly, the original exercise is inconsistent. 

https://github.com/fpinscala/fpinscala/blob/master/exercises/src/main/scala/fpinscala/parallelism/Nonblocking.scala#L152